### PR TITLE
Natrual iOS Viewport Binding

### PR DESCRIPTION
**Pull request by jridgewell**
An update to your `ViewportBindingNaturalIosEmbed` code to allow scrolling over Flying Carpets. Also removes around 100 lines of code by having the ViewportBindings extend from a common base class.

Please review https://github.com/ampproject/amphtml/pull/3234/commits/cb5522a11f12f8dad7fdf91ab87e2fed0b6d0d36 for only the iOS changes.

### CODE REVIEW

File: css/amp.css [hunk 1]
```diff
@@ -27,7 +27,7 @@ html {
   overflow-x: hidden !important;
 }
 html, body {
-  height: auto !important;
+  height: 100% !important;
```
Comment (dvoytenko) [inline css/amp.css]: This was added to fix https://github.com/ampproject/amphtml/issues/2187. Is this resolved now?

**Comment by dvoytenko** [comment]
@jridgewell These are some very heavy changes. The existing structure translates very well into experimentation. I recommend we create a new viewport implementation and trigger it by experiment to allow us to test it sufficiently. All those CSS properties are loaded with issues.

I'd be happy to review a separate PR that would do the suggested refactoring if that's more convenient. Combining them together makes it very hard to review anyway.

File: src/service/viewport-impl.js [hunk 2]
```diff
@@ -794,42 +774,28 @@ export class ViewportBindingNaturalIosEmbed_ {
...
-    //   -webkit-overflow-scrolling: touch;
+    //   overflow-x: hidden;
     // }
     setStyles(documentElement, {
-      overflowY: 'auto',
       webkitOverflowScrolling: 'touch',
     });
     setStyles(documentBody, {
       overflowX: 'hidden',
-      overflowY: 'auto',
```
Comment (dvoytenko) [inline src/service/viewport-impl.js]: Does this remove the need in fixed-transfer-layer?

File: src/service/viewport-impl.js [hunk 3]
```diff
@@ -794,42 +774,28 @@ export class ViewportBindingNaturalIosEmbed_ {
...
-      right: 0,
-      bottom: 0,
     });
 
     // Insert scrollPos element into DOM. See {@link onScrolled_} for why
     // this is needed.
     this.scrollPosEl_ = this.win.document.createElement('div');
     this.scrollPosEl_.id = '-amp-scrollpos';
     setStyles(this.scrollPosEl_, {
-      position: 'absolute',
```
Comment (dvoytenko) [inline src/service/viewport-impl.js]: Why are these removed? This looks to assume a lot about top-level body layout, such as that padding will always be zero or that no any kind of reordering would be done. And w/o this - does this have a benefit?

File: src/service/viewport-impl.js [hunk 4]
```diff
@@ -934,16 +861,20 @@ export class ViewportBindingNaturalIosEmbed_ {
...
-  setScrollTop(scrollTop) {
-    this.setScrollPos_(scrollTop || 1);
+  setScrollTop(scrollPos) {
+    // Stop us from scrolling _all_ the way to the top and triggering an
+    // annoying bounce animation.
+    scrollPos = Math.max(scrollPos, 1);
+    if (!this.scrollMoveEl_) {
+      return;
+    }
+    scrollPos -= this.getScrollTop();
```
Comment (dvoytenko) [inline src/service/viewport-impl.js]: This one is not quiet clear. Why do we have a relative expression when setting an absolute value?

**Comment by dvoytenko** [comment]
Otherwise, on first run through - everything looks great. But would be nice to get it into PROD in some form for a wider real-life document testing.

File: src/service/viewport-impl.js (hunk 2 above)
Comment (jridgewell) [inline src/service/viewport-impl.js]: No, it's still needed. 😞

File: src/service/viewport-impl.js (hunk 3 above)
Comment (jridgewell) [inline src/service/viewport-impl.js]: Removing the `position: absolute` on the `body` forced this. Because it's no longer positioned, any child that _is_ now positions itself relative to the viewport. So if we have a `position: absolute; top: 0; left: 0`, it will actually always be in the top-left corner of the screen.

> such as that padding will always be zero

Should we enforce that?

> any kind of reordering would be done

That's the same requirement as `endPosEl_`, so I didn't think it'd be an issue.

File: src/service/viewport-impl.js (hunk 4 above)
Comment (jridgewell) [inline src/service/viewport-impl.js]: See my [reply](https://github.com/ampproject/amphtml/pull/3234#discussion_r63443828). It's due to `scrollMoveEl_` being absolutely positioned relative to the viewport, not the scrolling body. So, we need to compensate for the body's scroll position to get it to the right spot.

File: src/service/viewport-impl.js (hunk 3 above)
Comment (dvoytenko) [inline src/service/viewport-impl.js]: So, what would happen to an element with user's CSS with `position: absolute; top: 0`? Where would it be laid out?

File: css/amp.css (hunk 1 above)
Comment (jridgewell) [inline css/amp.css]: No idea, will have to test. But this is the only way to trigger iOS scrolling with the other css changes I made.

File: src/service/viewport-impl.js (hunk 3 above)
Comment (jridgewell) [inline src/service/viewport-impl.js]: It behaves like a `position: fixed; top: 0`. We could fix this by making `body` `position: relative`. In fact, I'm not sure why I didn't do that in the first place.

Comment (dvoytenko) [inline src/service/viewport-impl.js]: Yes, `body` definition then have `position: relative` since we allow free use of `position: absolute` and it'd be too expensive to override all of those styles.

File: src/service/viewport-impl.js (hunk 4 above)
Comment (jridgewell) [inline src/service/viewport-impl.js]: No longer necessary since `body` is now `position: relative`.

File: src/service/viewport-impl.js (hunk 3 above)
Comment (jridgewell) [inline src/service/viewport-impl.js]: Fixed.

File: src/service/viewport-impl.js (hunk 4 above)
Comment (dvoytenko) [inline src/service/viewport-impl.js]: Awesome! That makes it a lot better.

**Comment by jridgewell** [comment]
As the CSS in this PR fixes https://github.com/ampproject/amphtml/pull/3234, I can pear this down to just the CSS changes and move the larger `ViewportBiding` refactors into another PR.

**Comment by dvoytenko** [comment]
@jridgewell The CSS part is actually the part that concerns me the most :) I'll just jump in and make a temporary fix for this on API level which we will remove as soon as we launch the updated solution.
